src/ring/logger.clj

```diff
@@ -112,33 +144,25 @@
                   transform-fn identity
                   log-exceptions? true
                   request-keys default-request-keys}}]
-   (fn [request]
-     (let [start-ms (or (::start-ms request)
-                        (System/currentTimeMillis))
-           log (make-transform-and-log-fn transform-fn log-fn)
-           base-message (select-keys request request-keys)]
-       (try
-         (let [{:keys [status] :as response} (handler request)
-               elapsed-ms (- (System/currentTimeMillis) start-ms)
-               level (if (and (number? status)
-                              (<= 500 status))
-                       :error
-                       :info)]
-           (log {:level level
-                 :message (-> base-message
-                              (assoc ::type :finish
-                                     :status status
-                                     ::ms elapsed-ms))})
-           response)
-         (catch Exception e
-           (when log-exceptions?
-             (let [elapsed-ms (- (System/currentTimeMillis) start-ms)]
-               (log {:level :error
-                     :throwable e
-                     :message (-> base-message
-                                  (assoc ::type :exception
-                                         ::ms elapsed-ms))})))
-           (throw e)))))))
+   (fn
+     ([request]
+      (let [start-ms (or (::start-ms request)
+                         (System/currentTimeMillis))
+            log (make-transform-and-log-fn transform-fn log-fn)
+            base-message (select-keys request request-keys)]
+        (try
+          (log-response (handler request) start-ms log base-message)
+          (catch Exception ex
+            (when log-exceptions? (log-exception ex start-ms log base-message))
+            (throw ex)))))
+     ([request respond raise]
+      (let [start-ms (or (::start-ms request)
+                         (System/currentTimeMillis))
+            log (make-transform-and-log-fn transform-fn log-fn)
+            base-message (select-keys request request-keys)]
+        (handler request
```

Comment:
@Kauko what do you think about wrapping this in try..catch and logging the exception?

Reply:
I had a try-catch here at first, but I realised async handlers use the raise function to handle exceptions, so the try-catch doesn't do anything.